### PR TITLE
[33] Produce mean gender differences with subject specific terminology removed

### LINKED ISSUE #33: Produce mean gender differences with subject specific terminology removed

### PR DESCRIPTION
**Pull request by Jack-0-0**
Closes #33.

BIT has defined two groups of subject specific words in the GCSE descriptions, 'crucial' and 'optional'.

This PR adds functionality to produce mean gender difference charts with the above words excluded:

* Adds subject specific word lists to inputs (`inputs/data/subject_specific_terminology/`)
* Adds getter for subject specific words lists (`getters/subject_terminology.py`)
* Updates `pipeline/glove_differences/make_mean_differences.py` to produce mean gender difference calculations with subject specific words removed 
* Updates `getters/mean_gender_differences.py` to be able to load mean gender difference csv files in the new format
* Updates `analysis/save_charts_data.py` to produce charts with subject specific words removed
* Replace old charts with new charts

---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

### CODE REVIEW

File: comp_sci_gender_bias/getters/subject_terminology.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    )
+
+
+def subject_specific_words(subject: str, specific_word_type: str) -> list:
+    """Create list of subject specific words
+
+    Args:
+        subject: "geo" or "cs"
+        specific_word_type: "crucial" or "optional"
+
```
Comment (georgerichardson) [inline comp_sci_gender_bias/getters/subject_terminology.py:29]:
```suggestion
```

File: comp_sci_gender_bias/getters/subject_terminology.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    Returns:
+        List of subject specific words
+    """
+    cat_words = categorised_words(subject)
+    return cat_words.query(f"`{specific_word_type} subject specific word` == 1")[
+        "word"
+    ].to_list()
```
Comment (georgerichardson) [inline comp_sci_gender_bias/getters/subject_terminology.py:36]: Minor stylistic
```suggestion
    return (
        cat_words
        .query(f"`{specific_word_type} subject specific word` == 1")["word"]
        .to_list()
    )
```

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py [hunk 3]
```diff
@@ -6,9 +6,11 @@
 )
 from comp_sci_gender_bias.getters.school_data import text_descriptions
 from comp_sci_gender_bias.getters.scraped_data import scraped_data
+from comp_sci_gender_bias.getters.subject_terminology import subject_specific_words
 import pandas as pd
 from comp_sci_gender_bias import PROJECT_DIR
 from comp_sci_gender_bias.utils.io import make_path_if_not_exist
+from typing import Union
```
Comment (georgerichardson) [inline comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py]: Minor point on readability - I try to order imports alphabetically based on the package name and, if there are many, to group them into chunks using one newline separator e.g. external libraries at the top, imports from the current module below and typing below that. Other people probably have better systems though!

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py [hunk 4]
```diff
@@ -31,33 +34,81 @@ def calc_mean_gender_diff(
...
             "POS": POS_LABELS,
             "mean_gender_diff": [
                 sub_word_pos_corpus.query(query)["Male - Female"].mean()
                 for query in POS_QUERIES
             ],
-            "subject": [subject_lbl] * len(POS_LABELS),
-            "data_source": [data_source_lbl] * len(POS_LABELS),
+            "subject": [subject] * row_length,
+            "data_source": [data_source_lbl] * row_length,
+            "words_removed": [word_removal] * row_length,
```
Comment (georgerichardson) [inline comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py]: pandas will expand scalar values to fill the length of the dataframe on construction
```suggestion
            "subject": subject,
            "data_source": data_source_lbl,
            "words_removed": word_removal,
```

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py [hunk 5]
```diff
@@ -31,33 +34,81 @@ def calc_mean_gender_diff(
...
+
+    words = [
+        word
+        for word in sub_word_pos_corpus["Word"].tolist()
+        if word not in words_to_remove
+    ]
+    glove_sims = glove_dists.gender_similarity_difference_word_list(words)
     sub_word_pos_corpus["Male - Female"] = sub_word_pos_corpus["Word"].map(glove_sims)
     sub_word_pos_corpus = sub_word_pos_corpus.dropna()
+    row_length = len(POS_LABELS)
```
Comment (georgerichardson) [inline comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py]:
```suggestion
```

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py [hunk 6]
```diff
@@ -31,33 +34,81 @@ def calc_mean_gender_diff(
...
     return pd.DataFrame.from_dict(
         {
             "POS": POS_LABELS,
             "mean_gender_diff": [
                 sub_word_pos_corpus.query(query)["Male - Female"].mean()
                 for query in POS_QUERIES
             ],
```
Comment (georgerichardson) [inline comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py]: If it was me, I would probably define this before constructing the dataframe to make the dict readable, but it's fine as is

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py [hunk 7]
```diff
@@ -31,33 +34,81 @@ def calc_mean_gender_diff(
...
+    words = [
+        word
+        for word in sub_word_pos_corpus["Word"].tolist()
+        if word not in words_to_remove
+    ]
+    glove_sims = glove_dists.gender_similarity_difference_word_list(words)
     sub_word_pos_corpus["Male - Female"] = sub_word_pos_corpus["Word"].map(glove_sims)
     sub_word_pos_corpus = sub_word_pos_corpus.dropna()
+    row_length = len(POS_LABELS)
     return pd.DataFrame.from_dict(
```
Comment (georgerichardson) [inline comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py:71]: n.b. you can just pass dicts in when constructing a dataframe like `pd.DataFrame({"a", [1, 2, 3], "b": 4})` and it knows what to do although I think there's an argument for being explicit too

**Review by georgerichardson — COMMENTED** [review]
Looks good. It might be good to add/update a readme explaining the difference between crucial, optional and no word removal.

File: comp_sci_gender_bias/getters/subject_terminology.py (hunk 2 above)
Comment (Jack-0-0) [reply comp_sci_gender_bias/getters/subject_terminology.py:36]: I prefer your suggestion too but black auto formats it to the way it is here.

File: comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py (hunk 4 above)
Comment (Jack-0-0) [reply comp_sci_gender_bias/pipeline/glove_differences/make_mean_differences.py]: Ahh, I didn't realise, great!

**Comment by Jack-0-0** [comment]
> Looks good. It might be good to add/update a readme explaining the difference between crucial, optional and no word removal.

Thanks for the review. I've made your suggested changes and added to the readme.

**Pull request merged**